### ISSUE TITLE
Adams2019 autoscheduler can produce schedules that aren't bounds-resilient

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#5070]
In the wild, we saw an autogenerated schedule that had estimates input and output edge sizes of 4096; one of the resulting schedules had something like this as the result:

```
    result.split(x, x, xi, 1008, TailStrategy::ShiftInwards)
        .split(y, y, yi, 378, TailStrategy::ShiftInwards)
        .split(yi, yi, yii, 6, TailStrategy::ShiftInwards)
        .split(yii, yii, yiii, 2, TailStrategy::ShiftInwards)
        .split(xi, xi, xii, 16, TailStrategy::ShiftInwards)
        .vectorize(xii)
        .compute_root()
        .reorder(xii, xi, yiii, yii, yi, x, y)
        .fuse(x, y, x)
        .parallel(x);
```

This effectively requires that the `result` buffer must be >= 1008 wide and >= 378 high (otherwise an assertion failure will result); this surely isn't desirable. We should likely take pains to ensure that splits for input and output functions all use GuardWithIf by default.

**Comment by abadams** [comment]
It's not great, but this was an intentional choice. The ShiftInwards splits can be manually changed to GuardWithIf if the pipeline must work with sizes smaller than the estimates, but performance will suffer, so it's a bad default. The schedules are only guaranteed to work for things at least as large as the estimate. Longer term we could have a way to specify minimum and maximum sizes in the estimates maybe?

**Comment by steven-johnson** [comment]
> It's not great, but this was an intentional choice. The ShiftInwards splits can be manually changed to GuardWithIf if the pipeline must work with sizes smaller than the estimates, but performance will suffer, so it's a bad default.

I strongly disagree here; in this case, the splits were different for each multitarget variants, with the largest (-> most-restrictive) only showing up for the base target. The result was that you'd only end up with runtime failures if you happened to run on a baseline (sse4.1-only) machine. Since autoscheduler output is targeted at the less-sophisticated user, and thus the user least likely to inspect the output for this sort of thing, we should err on the side of safety, not performance.

> The schedules are only guaranteed to work for things at least as large as the estimate. 

Is that explicitly documented anywhere? I don't recall seeing that promise anywhere outside of the code. And it's not the behavior I'd expect from the language ("estimate" doesn't imply "minimum" or "maximum"). 

> Longer term we could have a way to specify minimum and maximum sizes in the estimates maybe?

Maybe, though I can't recall ever seeing a case where "larger than expected" was a problem (except for the degenerate "needs > 32 bits case"). Assigning a minimum would be useful because then we could at least avoid various checks for (e.g.) always-wider-than-vector-width.

Maybe a good short-term option would be to add a flag to choose between "assume minimum = estimate" and "assume minimum = 1"? (Or "assume minimum = CONSTANT"?)

**Comment by abadams** [comment]
You make good points. GuardWithIf is problematic though. It's unfortunately still flaky when it comes to bounds inference being overconservative. Adding if statements is also a source of constant problems for llvm optimization passes. A lot of the progress we've made on making the gpu autoscheduler work a higher fraction of the time is by avoiding GuardWithIf whenever possible. An alternative way to implement assume minimum = 1 in a way that's going to work for sure would be adding a memcpy stage to the end of the pipeline that somehow only runs if the size would be too small. I wonder if we can reach that with our existing set of scheduling directives. I don't think we can. It could be done with a wrapper around the pipeline that does a bounds query...

It's not obvious to me how to implement "assume minimum = 1" well.

**Comment by benzwt** [comment]
As a beginner, I always use auto-scheduler.  
I use a quick and dirty way to *mitigate* this issue.

1. generate an object file(a.o) and its corresponding schedule file and header
2. change ShiftInwards to GuardWithIf for 1008 and 378
3. include the modified schedule file to the generator and generate an object file (b.o)
4. put a.o and b.o to a static library
4. add an interface before calling a and b
```
#include "a.h"
#include "b.h"

int interface(struct h_buffer_t *one, struct h_buffer_t *two)
{
    if (one->dim[0] < 1008 || one->dim[0] < 378)
    {
        return b(one, two);
    }
    else
    {
       return a(one, two);
    }
 }
```

**Comment by benzwt** [comment]
> As a beginner, I always use auto-scheduler.
> I use a quick and dirty way to _mitigate_ this issue.
> 
>     1. generate an object file(a.o) and its corresponding schedule file and header
> 
>     2. change ShiftInwards to GuardWithIf for 1008 and 378
> 
>     3. include the modified schedule file to the generator and generate an object file (b.o)
> 
>     4. put a.o and b.o to a static library
> 
>     5. add an interface before calling a and b
> 
> 
> ```
> #include "a.h"
> #include "b.h"
> 
> int interface(struct h_buffer_t *one, struct h_buffer_t *two)
> {
>     if (one->dim[0] < 1008 || one->dim[0] < 378)
>     {
>         return b(one, two);
>     }
>     else
>     {
>        return a(one, two);
>     }
>  }
> ```

I found that my solution just prevent the program from crash *but the result is totally wrong*.
I have to re-estimate the dimension of the image such as {{0,1008}, {0, 378}}, iteratively until the auto-scheduler produce *GuardWithIf*.

**Comment by benzwt** [comment]
> You make good points. GuardWithIf is problematic though. It's unfortunately still flaky when it comes to bounds inference being overconservative. Adding if statements is also a source of constant problems for llvm optimization passes. A lot of the progress we've made on making the gpu autoscheduler work a higher fraction of the time is by avoiding GuardWithIf whenever possible. An alternative way to implement assume minimum = 1 in a way that's going to work for sure would be adding a memcpy stage to the end of the pipeline that somehow only runs if the size would be too small. I wonder if we can reach that with our existing set of scheduling directives. I don't think we can. It could be done with a wrapper around the pipeline that does a bounds query...
> 
> It's not obvious to me how to implement "assume minimum = 1" well.

Can *specialize* come to rescue ?
I saw a simple test case about avoid bound issue using specialize (under test/). Somehow it is now missing.

**Comment by ryanstout** [comment]
Any updates to this issue? I've been doing Halide development for a few years now, but for some of my complex pipelines, Adams2021 beats anything I can do by hand. I realize GuardWithIf has a performance hit, but might be worth adding an ENV or something to let it use that when required. Thanks!
